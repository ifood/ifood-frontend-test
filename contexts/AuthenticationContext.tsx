import React, { createContext, useState, useEffect, useContext } from 'react';
import { useRouter } from 'next/router';
import qs from 'qs';
import Cookie from 'js-cookie';

type AuthContextType = {
  isAuthenticated: boolean;
};

const AuthContext = createContext<Partial<AuthContextType>>({});

export const AuthProvider = ({ children }) => {
  const [token, setToken] = useState(null);
  const router = useRouter();

  const path = router.asPath;

  useEffect(() => {
    const params = qs.parse(path);
    const authToken = params['/playlists#access_token'];
    setToken(authToken);
    Cookie.set('USER_TOKEN', authToken);
  }, [path]);

  return (
    <AuthContext.Provider value={{ isAuthenticated: !!token }}>
      {children}
    </AuthContext.Provider>
  );
};

export default function useAuth() {
  const context = useContext(AuthContext);

  return context;
}
