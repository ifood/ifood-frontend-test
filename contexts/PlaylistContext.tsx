import React, {
  createContext,
  useState,
  useMemo,
  useCallback,
  useEffect,
  useContext,
  useRef,
} from 'react';
import {
  FeaturedPlaylistsInterface,
  PlaylistInterface,
} from '../interfaces/playlist.interface';
import {
  getFeaturedPlaylists,
  getChoicesForFilter,
} from '../services/playlist.service';
import { FilterInterface } from '../interfaces/filter.interface';

interface PlaylistStateInterface {
  featuredPlaylists: FeaturedPlaylistsInterface | null;
  filterFields: FilterInterface[];
  filterChoices: FilterInterface | null;
  playlists: PlaylistInterface[];
}

type PlaylistContext = {
  state: PlaylistStateInterface;
  filterFields: FilterInterface[];
  filterChoices: FilterInterface;
  featuredPlaylists: FeaturedPlaylistsInterface;
  setFilterChoices: (filter) => any;
  removeFilterChoices: (id) => any;
  filterByText: (name: string, text: string) => any;
};

const REFRESH_RATE = 30000;

const INITIAL_STATE = {
  featuredPlaylists: null,
  filterFields: [],
  filterChoices: null,
  playlists: [],
};

const PlaylistContext = createContext<Partial<PlaylistContext>>({});

export const PlaylistProvider = contextProps => {
  const [state, setState] = useState<PlaylistStateInterface>(INITIAL_STATE);

  const value = useMemo(() => {
    return {
      setState,
      state,
    };
  }, [state]);

  const interval = useRef<number>();

  const changeState = (changedValue, keyValue) => {
    setState(prevState => ({
      ...prevState,
      [keyValue]: changedValue,
    }));
  };

  const fetchFeaturedPlaylists = useCallback(async () => {
    const params = state.filterChoices || {};
    try {
      const { data } = await getFeaturedPlaylists(params);

      if (data && Object.keys(data)) {
        changeState(data, 'featuredPlaylists');
        if (
          data.playlists &&
          data.playlists.items &&
          data.playlists.items.length
        ) {
          changeState(data.playlists.items, 'playlists');
        }
      }
    } catch (error) {
      console.log(error);
    }
  }, [state.filterChoices]);

  const fetchFilters = useCallback(async () => {
    try {
      const { data } = await getChoicesForFilter();
      if (data && Object.keys(data) && data.filters.length) {
        changeState([...data.filters], 'filterFields');
      }
    } catch (error) {
      console.log(error);
    }
  }, []);

  const setFilterChoices = useCallback(
    filter => {
      setState(prevState => ({
        ...prevState,
        filterChoices: { ...prevState.filterChoices, ...filter },
      }));
    },
    [setState],
  );

  const removeFilterChoices = useCallback(
    async id => {
      setState(prevState => {
        const nextState = prevState;
        delete nextState.filterChoices[id];
        return nextState;
      });

      await fetchFeaturedPlaylists();
    },
    [fetchFeaturedPlaylists],
  );

  const filterByText = useCallback(
    (name: string, text: string) => {
      const { featuredPlaylists } = state;
      const hasPlaylists =
        Object.keys(featuredPlaylists) &&
        featuredPlaylists.playlists &&
        featuredPlaylists.playlists.items &&
        featuredPlaylists.playlists.items.length;

      if (hasPlaylists && name === 'name') {
        const filteredByNames = featuredPlaylists.playlists.items.filter(
          playlist => {
            const replace = new RegExp(text.toLowerCase(), 'g');
            const hasMatch = replace.test(playlist.name.toLowerCase());
            return hasMatch;
          },
        );

        changeState(filteredByNames, 'playlists');
      }
    },
    [state],
  );

  useEffect(() => {
    const filterChoices = state.filterChoices || {};
    if (interval.current) {
      clearInterval(interval.current);
    }
    if (Object.keys(filterChoices).length) {
      const fetchWithParams = async () => {
        await fetchFeaturedPlaylists();
      };

      interval.current = setInterval(() => {
        fetchWithParams();
      }, REFRESH_RATE);
    }
  }, [fetchFeaturedPlaylists, state.filterChoices]);

  useEffect(() => {
    if (interval.current) {
      clearInterval(interval.current);
    }
    fetchFilters();
    fetchFeaturedPlaylists();
    interval.current = setInterval(() => {
      fetchFeaturedPlaylists();
    }, REFRESH_RATE);
  }, [fetchFeaturedPlaylists, fetchFilters]);

  return (
    <PlaylistContext.Provider
      value={{ ...value, setFilterChoices, removeFilterChoices, filterByText }}
      {...contextProps}
    />
  );
};

export default function usePlaylist() {
  const context = useContext(PlaylistContext);

  if (!context) {
    throw new Error('usePlaylist must go inside PlaylistProvider');
  }

  return {
    ...context.state,
    ...context,
  };
}
