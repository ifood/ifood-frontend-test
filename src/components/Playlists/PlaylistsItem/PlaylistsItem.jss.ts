import { makeStyles, Theme } from '@material-ui/core/styles';

export const useStyles = makeStyles(
  (theme: Theme) => ({
    image: {
      width: '100%',
      height: '100%',
      objectFit: 'cover',
      objectPosition: 'center center',
    },
    name: {
      fontWeight: theme.typography.fontWeightMedium,
      marginBottom: theme.spacing(1),
    },
    description: {
      display: '-webkit-box',
      boxOrient: 'vertical',
      lineClamp: 2,
      lineHeight: '1rem',
      height: '2rem',
      overflow: 'hidden',
      fontWeight: theme.typography.fontWeightMedium,
      '& > a': {
        color: theme.palette.text.primary,
        textDecoration: 'none',
        '&:hover': {
          textDecoration: 'underline',
        },
      },
    },
  }),
  { name: 'PlaylistsItem' }
);
