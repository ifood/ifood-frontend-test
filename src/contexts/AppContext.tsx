import React from 'react';
import { ReactQueryConfigProvider, ReactQueryConfig } from 'react-query';
import { MuiThemeProvider } from '@material-ui/core/styles';
import { MuiPickersUtilsProvider } from '@material-ui/pickers';
import CssBaseline from '@material-ui/core/CssBaseline';
import LuxonUtils from '@date-io/luxon';
import { theme } from '../theme';

const queryConfig: ReactQueryConfig = {
  queries: {
    refetchInterval: 30 * 1000,
    refetchIntervalInBackground: true,
    refetchOnWindowFocus: false,
  },
};

export const AppProvider: React.FC = ({ children }) => {
  return (
    <ReactQueryConfigProvider config={queryConfig}>
      <MuiThemeProvider theme={theme}>
        <MuiPickersUtilsProvider utils={LuxonUtils} locale="pt-BR">
          <CssBaseline />
          {children}
        </MuiPickersUtilsProvider>
      </MuiThemeProvider>
    </ReactQueryConfigProvider>
  );
};
