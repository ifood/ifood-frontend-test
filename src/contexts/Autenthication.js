import { createContext, useState, useEffect, useContext } from 'react'
import { element } from 'prop-types'
import { useRouter } from 'next/router'
import { setUserToken } from '../services/auth'
import qs from 'qs'

const AuthContext = createContext({})

export const AuthProvider = ({ children }) => {
  const [loading, setLoading] = useState(true)
  const [token, setToken] = useState(null)
  const router = useRouter()

  useEffect(() => {
    const path = router.asPath
    const queryParams = qs.parse(path)
    const accessToken = queryParams['/playlists#access_token']
    setToken(accessToken)
    setUserToken(accessToken)
    setLoading(false)
  }, [router])

  return (
    <AuthContext.Provider value={{ isAuthenticated: !!token, loading }}>
      {children}
    </AuthContext.Provider>
  )
}

export default function useAuth() {
  const context = useContext(AuthContext)

  return context
}

AuthProvider.propTypes = {
  children: element.isRequired
}
