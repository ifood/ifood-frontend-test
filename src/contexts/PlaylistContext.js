import {
  createContext,
  useState,
  useEffect,
  useContext,
  useMemo,
  useCallback
} from 'react'
import {
  getFeaturedPlaylists,
  getChoicesForFilter
} from '../services/playlist.service'
import { element } from 'prop-types'

const PlaylistContext = createContext({})
const INITIAL_STATE = {
  loadingPlaylists: true,
  loadingFilterFields: true,
  featuredPlaylists: {},
  filterFields: [],
  filterChoices: {}
}

export const PlaylistProvider = (contextProps) => {
  const [state, setState] = useState(INITIAL_STATE)

  const value = useMemo(() => {
    return {
      setState,
      state
    }
  }, [state])

  const featuredPlaylists = async () => {
    try {
      const { data } = await getFeaturedPlaylists()
      if (data && Object.keys(data)) {
        setState((prevState) => ({
          ...prevState,
          featuredPlaylists: data
        }))
      }
    } catch (error) {
      console.log(error)
    } finally {
      setState((prevState) => ({
        ...prevState,
        loadingPlaylists: false
      }))
    }
  }

  const choicesForFilter = async () => {
    try {
      const { data } = await getChoicesForFilter()
      if (data && Object.keys(data) && data.filters.length) {
        setState((prevState) => ({
          ...prevState,
          filterFields: data.filters
        }))
      }
    } catch (error) {
      console.log(error)
    } finally {
      setState((prevState) => ({
        ...prevState,
        loadingFilterFields: false
      }))
    }
  }

  useEffect(() => {
    const fetch = async () => {
      await featuredPlaylists()
      await choicesForFilter()
    }

    fetch()
  }, [])

  return <PlaylistContext.Provider value={value} {...contextProps} />
}

export default function usePlaylist() {
  const context = useContext(PlaylistContext)
  const { state, setState } = context
  if (!context) {
    throw new Error('usePlaylist must go inside PlaylistProvider')
  }

  const setFilterChoices = useCallback(
    (filter) => {
      setState((prevState) => ({
        ...prevState,
        filterChoices: { ...prevState.filterChoices, ...filter }
      }))
    },
    [setState]
  )

  return {
    ...state,
    setFilterChoices
  }
}

PlaylistProvider.propTypes = {
  children: element.isRequired
}
