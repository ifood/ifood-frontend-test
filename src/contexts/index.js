import React, { createContext, useContext, useState } from 'react'

import { apiAuth, api, apiMocky } from '~/services/api'
import { getSetAccess } from '~/services/auth'
import { FEATUREDPLAYLISTSERVER } from '~/config'
import { dateFormattedPattern } from '~/utils'

import moment from 'moment'

const queryString = require('query-string')

const PlaylistContext = createContext()

const PlaylistProvider = ({ children }) => {
  const [showFilters, setShowFilters] = useState(false)
  const [loading, setLoading] = useState(true)
  const [filtersList, setFiltersList] = useState([])
  const [featuredList, setFeaturedList] = useState()
  const [playlistSearchName, setPlaylistSearchName] = useState('')

  const [locale, setLocale] = useState('pt_BR')
  const [country, setCountry] = useState('BR')
  const [timestamp, setTimestamp] = useState(
    moment().format(dateFormattedPattern)
  )
  const [limit, setLimit] = useState(30)
  const [offset, setOffset] = useState(1)

  const modal = (type) => {
    setShowFilters(type)
  }

  const auth = async () => {
    const response = await apiAuth.post(
      '',

      queryString.stringify({
        response_type: 'token',
        grant_type: 'client_credentials',
      })
    )

    getSetAccess(response.data)
  }

  const getQueryFilter = (filters) => {
    const encodeURI = encodeURIComponent

    let queryFilter = []

    for (let filter in filters) {
      if (filters[filter] !== '') {
        queryFilter.push(encodeURI(filter) + '=' + encodeURI(filters[filter]))
      }
    }

    return queryFilter.length ? `?${queryFilter.join('&')}` : ''
  }

  const filterPlaylistName = (name) => {
    return (
      filtersList &&
      featuredList &&
      featuredList.filter((item) =>
        item.name.toLowerCase().includes(name.toLowerCase())
      )
    )
  }

  const getFeaturedList = async (filters) => {
    const queryFilter = getQueryFilter(filters)

    const response = await api.get(`${FEATUREDPLAYLISTSERVER}${queryFilter}`)

    setLoading(false)
    setFeaturedList(response.data.playlists.items)
  }

  const filtersUpdate = async (updatedFilter) => {
    setLocale(updatedFilter.locale)
    setCountry(updatedFilter.country)
    setTimestamp(updatedFilter.timestamp)
    setLimit(updatedFilter.limit)
    setOffset(updatedFilter.offset)

    getFeaturedList(updatedFilter)
  }

  const getFiltersList = async () => {
    const response = await apiMocky.get()

    setFiltersList(response.data.filters)
  }

  return (
    <PlaylistContext.Provider
      value={{
        auth,
        loading,
        showFilters,
        modal,
        getFiltersList,
        filtersList,
        filtersUpdate,
        filterPlaylistName,
        playlistSearchName,
        setPlaylistSearchName,
        getFeaturedList,
        featuredList,
        setFeaturedList,
        locale,
        country,
        timestamp,
        limit,
        offset,
        setLocale,
        setCountry,
        setTimestamp,
        setLimit,
        setOffset,
      }}
    >
      {children}
    </PlaylistContext.Provider>
  )
}

function useFeaturedMusic() {
  const context = useContext(PlaylistContext)

  if (!context) {
    throw new Error('Houve um erro, vocÃª precisa usar o PlaylistProvider')
  }

  return context
}

export { PlaylistProvider, useFeaturedMusic }
