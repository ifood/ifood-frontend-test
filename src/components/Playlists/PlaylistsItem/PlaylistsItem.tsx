import React from 'react';
import Card from '@material-ui/core/Card';
import CardActionArea from '@material-ui/core/CardActionArea';
import CardContent from '@material-ui/core/CardContent';
import CardMedia from '@material-ui/core/CardMedia';
import Typography from '@material-ui/core/Typography';
import { Playlist } from '../../../interfaces/playlist';
import { useStyles } from './PlaylistsItem.jss';

interface PlaylistsItemProps {
  playlist: Playlist;
}

export const PlaylistsItem: React.FC<PlaylistsItemProps> = ({
  playlist: { name, description, image, url },
}) => {
  const classes = useStyles();
  return (
    <Card>
      <CardActionArea href={url}>
        <CardMedia
          className={classes.image}
          image={image}
          component="img"
          title="Imagem playlist"
        />
        <CardContent>
          <Typography className={classes.name} noWrap>
            {name}
          </Typography>
          <Typography
            className={classes.description}
            variant="caption"
            color="textSecondary"
            dangerouslySetInnerHTML={{ __html: description }}
          />
        </CardContent>
      </CardActionArea>
    </Card>
  );
};
