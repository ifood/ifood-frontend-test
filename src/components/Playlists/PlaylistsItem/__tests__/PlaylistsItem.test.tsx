import React from 'react';
import { render, screen } from '@testing-library/react';
import { PlaylistsItem } from '../PlaylistsItem';
import { AppProvider } from '../../../../contexts/AppContext';

const playlist = {
  id: '1',
  name: 'Playlist name',
  description: 'Playlist description',
  image: 'image1',
  url: 'http://localhost/url',
};

describe('PlaylistsItem', () => {
  test('matches the snapshot', () => {
    const { baseElement } = render(<PlaylistsItem playlist={playlist} />, { wrapper: AppProvider });
    expect(baseElement).toMatchSnapshot();
  });

  test('renders the playlists item', () => {
    render(<PlaylistsItem playlist={playlist} />, { wrapper: AppProvider });

    const link = screen.getByRole('link') as HTMLAnchorElement;
    expect(link).toBeInTheDocument();
    expect(link.href).toEqual(playlist.url);

    expect(screen.getByRole('img')).toBeInTheDocument();

    expect(screen.getByText(/playlist name/i)).toBeInTheDocument();
    expect(screen.getByText(/playlist description/i)).toBeInTheDocument();
  });
});
