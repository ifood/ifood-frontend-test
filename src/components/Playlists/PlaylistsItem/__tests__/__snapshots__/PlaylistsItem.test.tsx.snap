// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PlaylistsItem matches the snapshot 1`] = `
<body>
  <div>
    <div
      class="MuiPaper-root MuiCard-root MuiPaper-elevation1 MuiPaper-rounded"
    >
      <a
        aria-disabled="false"
        class="MuiButtonBase-root MuiCardActionArea-root"
        href="http://localhost/url"
        tabindex="0"
      >
        <img
          class="MuiCardMedia-root PlaylistsItem-image-1 MuiCardMedia-media MuiCardMedia-img"
          src="image1"
          title="Imagem playlist"
        />
        <div
          class="MuiCardContent-root"
        >
          <p
            class="MuiTypography-root PlaylistsItem-name-2 MuiTypography-body1 MuiTypography-noWrap"
          >
            Playlist name
          </p>
          <span
            class="MuiTypography-root PlaylistsItem-description-3 MuiTypography-caption MuiTypography-colorTextSecondary"
          >
            Playlist description
          </span>
        </div>
        <span
          class="MuiCardActionArea-focusHighlight"
        />
        <span
          class="MuiTouchRipple-root"
        />
      </a>
    </div>
  </div>
</body>
`;
